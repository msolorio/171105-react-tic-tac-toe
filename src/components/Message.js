import React from 'react';
import PropTypes from 'prop-types';

export default function Message(props) {

  function getMessage() {
    switch(true) {
      // converts to a boolean value
      case !!props.winner:
        return `Winner: ${props.winner}`;
      case !props.squaresAvailable:
        return `Game over. There are no available squares.`;
      default:
        return `Player's turn: ${props.xIsNext ? 'X' : 'O'}`;
    }
  }

  return (
    <div className="Message">{getMessage()}</div>
  );
}

Message.propTypes = {
  winner: PropTypes.string.isRequired,
  squaresAvailable: PropTypes.bool.isRequired,
  xIsNext: PropTypes.bool.isRequired
};
