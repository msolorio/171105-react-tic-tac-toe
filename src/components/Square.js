import React from 'react';
import PropTypes from 'prop-types';

export default function Square(props) {

  function handleSquareClick() {
    props.onSquareClick(props.index);
  }

  return (
    <div className={`Square Square-${props.index}`}
      onClick={handleSquareClick}>
      {props.squareVal}
    </div>
  );
}

Square.propTypes = {
  index: PropTypes.number.isRequired,
  squareVal: PropTypes.string.isRequired,
  onSquareClick: PropTypes.func.isRequired
};
