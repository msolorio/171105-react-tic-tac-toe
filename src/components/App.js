import React, { Component } from 'react';
import Board from './Board';
import Message from './Message';

class App extends Component {
  constructor(props) {
    super(props);

    this.state = {
      squareVals: Array(9).fill(''),
      xIsNext: true
    };

    this.winningCombos = [
      [0, 1, 2],
      [3, 4, 5],
      [6, 7, 8],
      [0, 3, 6],
      [1, 4, 7],
      [2, 5, 8],
      [0, 4, 8],
      [6, 4, 2]
    ];

    this.handleSquareClick = this.handleSquareClick.bind(this);
  }

  calculateWinner() {
    const squareVals = this.state.squareVals;

    return this.winningCombos.reduce((winner, combo) => {
      const [a, b, c] = combo;

      if (
        squareVals[a] !== ''
        && squareVals[a] === squareVals[b]
        && squareVals[a] === squareVals[c]
      ) {
        return squareVals[a];
      }
      return winner;
    }, '');
  }

  // returns boolean value signifying if there are available squares left
  squaresAvailable() {
    return this.state.squareVals.reduce((squaresLeft, square) => {
      if (square === '') return true;
      return squaresLeft;
    }, false);
  }

  handleSquareClick(index) {
    if (this.calculateWinner()) return;
    if (this.state.squareVals[index] !== '') return;
    const squareValsClone = this.state.squareVals.slice();
    squareValsClone[index] = this.state.xIsNext ? 'X' : 'O';

    this.setState((prevState) => ({
      squareVals: squareValsClone,
      xIsNext: !prevState.xIsNext
    }));
  }

  render() {
    return (
      <div className="App">
        <Message winner={this.calculateWinner()}
          squaresAvailable={this.squaresAvailable()}
          xIsNext={this.state.xIsNext} />
        <Board squareVals={this.state.squareVals}
          onSquareClick={this.handleSquareClick} />
      </div>
    );
  }
}

export default App;
