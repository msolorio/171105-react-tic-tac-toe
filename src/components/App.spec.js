import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import ReactDOM from 'react-dom';
import App from './App';
import Square from './Square';
import Message from './Message';

describe('App', () => {
  it('should render without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
  });

  it('should match snapshot', () => {
    const tree = renderer.create(
      <App />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should render a message component', () => {
    const wrapper = shallow(<App />);
    expect(wrapper.find(Message).length).toBe(1);
  });

  it('should render a board component', () => {
    const wrapper = shallow(<App />);
    expect(wrapper.find(Message).length).toBe(1);
  });

});

/*************************************************************************
* TODO: revamp to use setState rather than manipulating state directly
**************************************************************************/
describe('handleSquareClick', () => {
  it('should update state when called when there is no winner and square is not occupied', () => {
    const wrapperInstance = shallow(<App />).instance();
    wrapperInstance.handleSquareClick(0);

    expect(wrapperInstance.state.squareVals[0]).toBe('X');
    expect(wrapperInstance.state.xIsNext).toBe(false);
  });

  it('should not update state when called on a square already occupied', () => {
    const wrapperInstance = shallow(<App />).instance();
    wrapperInstance.state.squareVals[0] = 'O';
    wrapperInstance.state.xIsNext = true;
    wrapperInstance.handleSquareClick(0);

    expect(wrapperInstance.state.squareVals[0]).toBe('O');
    expect(wrapperInstance.state.xIsNext).toBe(true);
  });

  it('should not update state if there is already a winner', () => {
    const winningCombos = shallow(<App />).instance().winningCombos;

    winningCombos.forEach((combo) => {
      const wrapperInstance = mount(<App />).instance();
      combo.forEach(position => wrapperInstance.state.squareVals[position] = 'O');
      wrapperInstance.state.xIsNext = true;
      wrapperInstance.handleSquareClick(combo[0]);

      expect(wrapperInstance.state.squareVals[combo[0]]).toBe('O');
      expect(wrapperInstance.state.xIsNext).toBe(true);
    });
  });
});

describe('calculateWinner', () => {

  it('should return empty string if there is no winner', () => {
    const wrapperInstance = shallow(<App />).instance();
    const winner = wrapperInstance.calculateWinner(wrapperInstance.state.squareVals);
    expect(winner).toBe('');
  });

  it('should return a winner if one exists in state', () => {
    const winningCombos = shallow(<App />).instance().winningCombos;

    winningCombos.forEach((combo) => {
      const wrapperInstance = mount(<App />).instance();
      combo.forEach(position => wrapperInstance.state.squareVals[position] = 'X');

      expect(wrapperInstance.calculateWinner()).toBe('X');
    });
  });
});

describe('squaresAvailable', () => {
  it('should return true when no squares have been clicked', () => {
    const wrapperInstance = shallow(<App />).instance();

    expect(wrapperInstance.squaresAvailable()).toBe(true);
  });

  it('should return false when all squares have been filled', () => {
    const wrapperInstance = shallow(<App />).instance();
    const filledSquareVals = ['X', 'O', 'X', 'O', 'X', 'O', 'X', 'O', 'X'];
    const squareVals = wrapperInstance.state.squareVals = filledSquareVals;

    expect(wrapperInstance.squaresAvailable()).toBe(false);
  });
});
