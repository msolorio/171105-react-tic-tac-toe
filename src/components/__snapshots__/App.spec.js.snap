// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `
<div
  className="App"
>
  <div
    className="Message"
  >
    Player's turn: X
  </div>
  <div
    className="Board"
  >
    <div
      className="Square Square-0"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="Square Square-1"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="Square Square-2"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="Square Square-3"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="Square Square-4"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="Square Square-5"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="Square Square-6"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="Square Square-7"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="Square Square-8"
      onClick={[Function]}
    >
      
    </div>
  </div>
</div>
`;
